lighthouse-core/audits/seo/robots-txt.js

```diff
@@ -0,0 +1,195 @@
+  }
+
+  line = line.trim();
+
+  if (line.length === 0) {
+    return {};
+  }
+
+  const colonIndex = line.indexOf(':');
+
+  if (colonIndex === -1) {
+    throw new Error('Syntax not understood');
+  }
+
+  const directiveName = line.slice(0, colonIndex).trim().toLowerCase();
+  const directiveValue = line.slice(colonIndex + 1).trim();
+
+  if (!DIRECTIVE_SAFELIST.has(directiveName)) {
+    throw new Error('Unknown directive');
+  }
+
+  if (directiveName === DIRECTIVE_SITEMAP) {
+    let sitemapUrl;
+
+    try {
+      sitemapUrl = new URL(directiveValue);
+    } catch (e) {
+      throw new Error('Invalid sitemap URL');
+    }
+
+    if (!SITEMAP_VALID_PROTOCOLS.has(sitemapUrl.protocol)) {
+      throw new Error('Invalid sitemap URL protocol');
+    }
+  }
+
+  if (directiveName === DIRECTIVE_USER_AGENT && !directiveValue) {
+    throw new Error('No user-agent specified');
+  }
+
+  if (directiveName === DIRECTIVE_ALLOW || directiveName === DIRECTIVE_DISALLOW) {
+    if (directiveValue !== '' && directiveValue[0] !== '/' && directiveValue[0] !== '*') {
+      throw new Error('Pattern should either be empty, start with "/" or "*"');
+    }
+
+    const dolarIndex = directiveValue.indexOf('$');
+
+    if (dolarIndex !== -1 && dolarIndex !== directiveValue.length - 1) {
+      throw new Error('"$" should only be used at the end of the pattern');
+    }
+  }
+
+  return {
+    directive: directiveName,
+    value: directiveValue,
+  };
+}
+
+function validateRobots(content) {
+  let inGroup = false;
+
+  return content
+    .split(/\r\n|\r|\n/)
+    .map((line, index) => {
+      let parsedLine;
+
+      try {
+        parsedLine = parseLine(line);
+      } catch (e) {
+        return {
+          index: index + 1,
+          line: line,
+          error: e.message,
+        };
+      }
+
+      if (parsedLine.directive === DIRECTIVE_USER_AGENT) {
+        inGroup = true;
+      } else if (!inGroup && DIRECTIVES_GROUP_MEMBERS.has(parsedLine.directive)) {
+        return {
+          index: index + 1,
+          line: line,
+          error: 'No user-agent specified',
+        };
+      }
+
+      return null;
+    })
+    .filter(error => error !== null);
+}
+
+class RobotsTxt extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      name: 'robots-txt',
+      description: 'robots.txt is valid',
+      failureDescription: 'robots.txt is not valid',
+      helpText: 'If your robots.txt file is malformed, crawlers may not be able to understand ' +
+      'how you want your website to be crawled or indexed.',
+      requiredArtifacts: ['RobotsTxt'],
+    };
+  }
+
+  /**
+   * @param {!Artifacts} artifacts
+   * @return {!AuditResult}
+   */
+  static audit(artifacts) {
+    const {
+      status,
+      content,
+    } = artifacts.RobotsTxt;
+
+    if (!status) {
+      return {
+        rawValue: false,
+        debugString: 'Lighthouse was unable to download your robots.txt file',
```

Comment:
should we mark this as not applicable instead?

Reply:
This will only happen when a `fetch` call fails which, AFAIK, can only happen on a network failure (or a CORS failure, but we are fetching a local resource here). IMO that's an extraordinary situation which prevents us from finishing the audit, ergo the error. Let me know if that makes sense to you!

Reply:
that's fair, sounds fine to me 👍